crates/pallet-domains/src/lib.rs

```diff
@@ -465,6 +517,62 @@ mod pallet {
             Ok(())
         }
 
+        #[pallet::call_index(7)]
+        #[pallet::weight(Weight::from_all(10_000))]
+        // TODO: proper benchmark
+        pub fn submit_bundle_v2(
+            origin: OriginFor<T>,
+            opaque_bundle: OpaqueBundleOf<T>,
+        ) -> DispatchResult {
+            ensure_none(origin)?;
+
+            log::trace!(target: "runtime::domains", "Processing bundle: {opaque_bundle:?}");
+
+            let domain_id = opaque_bundle.domain_id();
+            let bundle_hash = opaque_bundle.hash();
+            let extrinsics_root = opaque_bundle.extrinsics_root();
+            let operator_id = opaque_bundle.operator_id();
+            let (bundle_author, receipt) = opaque_bundle.deconstruct();
+
+            let receipt_type = execution_receipt_type::<T>(domain_id, &receipt);
+            if receipt_type == ReceiptType::Stale {
+                // The stale receipt should not be further processed, but we still track them for purposes
+                // of measuring the bundle production rate.
+                Self::note_domain_bundle(domain_id);
+                return Ok(());
+            }
+
+            // Add the exeuctione receipt to the block tree
+            process_execution_receipt::<T>(domain_id, operator_id, receipt, receipt_type)
+                .map_err(Error::<T>::from)?;
+
+            // `SuccessfulBundles` is empty means this is the first accepted bundle for this domain in this
+            // consensus block, which also mean a domain block will be produced thus update `HeadDomainNumber`
+            // to this domain block's block number.
+            if SuccessfulBundles::<T>::get(domain_id).is_empty() {
+                let next_number = HeadDomainNumber::<T>::get(domain_id)
+                    .checked_add(&One::one())
+                    .ok_or::<Error<T>>(BlockTreeError::MaxHeadDomainNumber.into())?;
+                HeadDomainNumber::<T>::set(domain_id, next_number);
+            }
+
+            // Put the `extrinsics_root` to the inbox of the current under building domain block
+            let head_domain_number = HeadDomainNumber::<T>::get(domain_id);
+            ExecutionInbox::<T>::append(domain_id, head_domain_number, extrinsics_root);
```

Comment:
Why not use `consensus_block_number` here, i.e., `ExecutionInbox = (domain_id, consensus_block_number, vec![bundle_extrinsics_root])`? Every domain block number has a corresponding consensus_block_number and it's already in the receipt. And then `HeadDomainNumber` is needless.

Reply:
I found using `consensus_block_number` as the key of `ExecutionInbox` will bring some complexity to its removal. 

Namely, when a new head receipt is added to block tree, the oldest height of the block tree will be removed but if there are multiple ERs exist in that height and each ER points to different consensus blocks then it is not easy to tell which consensus block's `ExecutionInbox` to remove as if we remove the wrong one it will make valid ER of that consensus block unverifiable.

After considering serval edge cases I have come out with a solution that may fix this issue, namely to remove the `ExecutionInbox` of the consensus block with the smallest block number and if it is already removed then move to the second small block and so on. I can explain how the solution came out but as you can see it is more complicated and not straightforward compare to the current approach.

Reply:
>each ER points to different consensus blocks

Actually, I think that is exactly another argument to use `consensus_block_number`, they shouldn't point to different consensus blocks. Given it's possible that there are multiple consensus blocks that could derive the same domain block when some of them may contain no bundles, let's say $Block_{C1}$ => $Block_{D1}$, $Block_{C2}$ => $Block_{D1}$,  ideally we should see $Receipt( C1, D1)$, but if someone submits a receipt $Receipt(C2, D1)$, when using `domain_block_number`, it will still be considered as valid if all the other fields are valid. If we use `consensus_block_number`, $Receipt(C2, D1)$ will be rejected as the execution inbox is not found. In other words, there is only one canonical receipt for a specific domain block, but there will be multiple using `domain_block_number` in theory, which is not ideal.

Reply:
> If we use consensus_block_number, $Receipt(C2, D1)$ will be rejected as the execution inbox is not found

If $Block_{C2}$ also contains bundles then using `consensus_block_number` will also considered $Receipt(C2, D1)$ as valid. I think there are cases that can't be defended by using either `domain_block_number` or `consensus_block_number` and will have to involve fraud proof.

My main concern about using `consensus_block_number` is the removal of execution inbox. When there is a new head receipt added to the block tree, the ERs with domain block number `head_receipt_number - BlockTreePruningDepth` will be removed, when using `domain_block_number` we can know exactly which execution inbox to remove but when using `consensus_block_number` it is a lot more complicated.

Reply:
What I saying was the case that both $Block_{C1}$ and $Block_{C2}$ contain bundles and we expected $Receipt(C1, D1)$ and $Receipt(C2, D2)$, but using `consensus_block_number` can still considered $Receipt(C2, D1)$ as valid.

Reply:
Well, that makes sense, then I think we need a pair of `(consensus_block_number, domain_block_number)` to obtain a canonical receipt for a domain block.